Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.Common/Providers/GraphUtility.cs

```diff
@@ -237,34 +198,56 @@ public async Task<HttpResponseMessage> ShareSchedule(string accessToken, string
             return shiftTeams;
         }
 
-        /// <summary>
-        /// Method that will obtain the Graph token.
-        /// </summary>
-        /// <param name="tenantId">The Tenant ID.</param>
-        /// <param name="instance">The instance.</param>
-        /// <param name="clientId">The App ID of the Web Application.</param>
-        /// <param name="clientSecret">The client secret.</param>
-        /// <param name="userId">The AAD Object ID of the Admin, could also be the UPN.</param>
-        /// <returns>A string that represents the Microsoft Graph API token.</returns>
-        public async Task<string> GetAccessTokenAsync(
-            string tenantId,
-            string instance,
-            string clientId,
-            string clientSecret,
-            string userId = default(string))
+        /// <inheritdoc/>
+        public async Task<HttpResponseMessage> SendGraphHttpRequest(GraphConfigurationDetails graphConfigurationDetails, HttpClient httpClient, HttpMethod httpMethod, string requestUrl, string requestString = null)
         {
-            string authority = $"{instance}{tenantId}";
-            var cache = new RedisTokenCache(this.cache, clientId);
+            using (var httpRequestMessage = new HttpRequestMessage(httpMethod, requestUrl))
+            {
+                if (!string.IsNullOrEmpty(requestString))
+                {
+                    httpRequestMessage.Content = new StringContent(requestString, Encoding.UTF8, "application/json");
+                }
+
+                var response = await httpClient.SendAsync(httpRequestMessage).ConfigureAwait(false);
+
+                if (response.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    // Refresh the access token and recall
+                    var newToken = await this.GetAccessTokenAsync(graphConfigurationDetails).ConfigureAwait(false);
+
+                    graphConfigurationDetails.ShiftsAccessToken = newToken;
+
+                    using (var retryRequestMessage = new HttpRequestMessage(httpMethod, requestUrl))
```

Comment:
Why are you creating a new message instead of updating the bearer token and retrying with the original message? If you create a new message like this, you will lose the other headers that the callers may have added such as the passthrough header.

Reply:
Discussed.